packages/driver/src/multidomain/index.js

```diff
@@ -80,11 +81,70 @@ const setup = () => {
   Cypress.on('log:added', onLogAdded)
   Cypress.on('log:changed', onLogChanged)
 
-  specBridgeCommunicator.on('run:domain:fn', (data) => {
-    // TODO: await this if it's a promise, or do whatever cy.then does
-    window.eval(`(${data.fn})()`)
+  specBridgeCommunicator.on('run:domain:fn', async ({ fn, isDoneFnAvailable = false }) => {
+    const deferredSwitchToDomain = createDeferred()
 
-    specBridgeCommunicator.toPrimary('run:domain:fn')
+    cy.state('switchToDomainDeferred', deferredSwitchToDomain)
+    const evalFn = `(${fn})()`
+
+    // await the eval func, whether it is a promise or not
+    // we should not need to transpile this as our target browsers support async/await
+    // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function for more details
+    const asyncWrapper = `(async () => {
+      const deferredSwitchToDomain = cy.state('switchToDomainDeferred')
+
+      try {
+        await ${evalFn}
+        deferredSwitchToDomain.resolve()
+      } catch(e){
+        deferredSwitchToDomain.reject(e)
+      }
+    })()`
```

Comment:
I think this can be cleaned up to not use a deferred. `eval` returns the result of the evaled function, so you can do a try/catch on the outside. Something like:

```js
const evalFn = `(${fn})()`

const asyncWrapper = `(async () => {
  return ${evalFn}
})()`

// later

try {
  window.eval(asyncWrapper)
} catch (err) {
  // ...
}
```

We also need to try/catch the done wrapper, because the user's fn could synchronously throw before done is invoked.

Reply:
> I think this can be cleaned up to not use a deferred. `eval` returns the result of the evaled function, so you can do a try/catch on the outside. Something like:
> 
> ```js
> const evalFn = `(${fn})()`
> 
> const asyncWrapper = `(async () => {
>   return ${evalFn}
> })()`
> 
> // later
> 
> try {
>   window.eval(asyncWrapper)
> } catch (err) {
>   // ...
> }
> ```
Oh cool I didn't know that (the more you know 🌈 )! I was able to refactor out the deferred promises so we should now get the same benefit without all the code 🚀 .

> We also need to try/catch the done wrapper, because the user's fn could synchronously throw before done is invoked.

I think this use case should now be handled by the refactor. Are you referring to something like this?

```js
  it('t', (done) => {
    // @ts-ignore
    cy.switchToDomain('foobar.com', () => {
      throw 'oops'
      done()
    })
  })
```

![](https://user-images.githubusercontent.com/3980464/148283516-e2b9a839-bdad-4c11-b3a7-1e96063dde53.png)

Reply:
Yeah exactly. It would be great if you could add a test for that.

Reply:
added in d04e4962721dd77866534c8db7a83e00fa34dadb